### ISSUE TITLE
optimization parameters for Blogcatalog node classification

### ISSUE THREAD

**Issue by agoodweathercc** [palash1992/GEM#57]
Hello, 

Here(https://github.com/palash1992/GEM/issues/52) you post the hyperparameters for blogcatalog node classification. However, for SDNE, the optimization of autoencoder involves quite some hyperparameter such as the number of iteration, batch size, learning rate...

I tried a few combinations but even running one epoch of SDNE for blogcatalog takes about 500s, I am not able to try too many hyperparameter combinations. And my result got so far is much worse than the result reported in the paper. (For other methods, I am able to replicate the results. SDNE seems particularly hard.) Would it be possible for you to share other hyperparameters(number of iteration, batch size, learning rate)?

Many Thanks!

**Comment by agoodweathercc** [comment]
btw, for those who are also having trouble using SDNE, I contacted the author of SDNE paper and haven't got any response. There are few Github repos implementing SDNE but none of them is able to replicate the result according to the authors!

**Comment by agoodweathercc** [comment]
sorry, I saw you post your n_iter and n_batch. The parameter that is still missing is embedding dimension.

also, nu1 and nu2 seem to me doesn't appear in the original paper. (there is a parameter v for L_reg in formula 5 of SDNE paper, I didn't see in your implementation which parameter correspond to that v).

Also, in the hyperparameter range you provide, alpha is 1e-5, which basically means that first order relationship is ignored. This is a bit surprising to me.

**Comment by palash1992** [comment]
The embedding dimension is missing because in the paper I test on a range of embedding dimensions and show results of all of them. If you want to try 1, 128 worked pretty well for this data set. I used both l1 and l2 regularization and nu1 and nu2 correspond to their coefficients. If you use u1=0 and u2=v it becomes the original SDNE.

And yes, I was also quite surprised to see that low values of alpha gives the best performance. But from the experiments it was very evident.

**Comment by agoodweathercc** [comment]
> The embedding dimension is missing because in the paper I test on a range of embedding dimensions and show results of all of them. If you want to try 1, 128 worked pretty well for this data set. I used both l1 and l2 regularization and nu1 and nu2 correspond to their coefficients. If you use u1=0 and u2=v it becomes the original SDNE.
> 
> And yes, I was also quite surprised to see that low values of alpha gives the best performance. But from the experiments it was very evident.

**Comment by agoodweathercc** [comment]
Thanks a lot for quick response. I am going to take your advice and use 128 as embedding dimension.
For 'beta': [2, 5, 10, 20], 'n_units': [[], [128], [512, 128], [1024, 512, 128]], there are still 16 combinations. Can you recall which particlular beta and n_units gives the best performance?

Also n_iter=100, but I tried some experiments and the loss is almost the same after ten epoches, is it really necessary to run 100 iterations? After all, even using two GPU, It still takes me 500s to run each iteration. This is way too slow for hyperparameter search.

**Comment by agoodweathercc** [comment]
It's really hard to replicate SDNE results and the right choice of hyperparameter is crucial. I appreciate that if you can further narrow down search space for blogcatalog.

**Comment by palash1992** [comment]
Please try beta=10 and n_units=[512, 128] and [1024, 512, 128]. Yes, you can lower the number of iterations. I kept them high because for some configurations of hyperparameters it took a long time to converge.

**Comment by agoodweathercc** [comment]
Thanks a lot. I use beta=5, n_units = [512,128], d=100 and finally is able to reproduce the same result in your paper. But it seems that the original paper reports better result than yours. For example, the original paper reports ~0.2-0.23 Macro-F1 while in your paper Macro-F1 for blogcatalog classification about 0.13.

Another question is about the understanding of n_units=[512, 128]. You say embedding dimension should be 128 but according to the sdne.py document,  n_units is a vector of length K-1 containing #units in hidden layers of encoder/decoder, not including the units in the embedding layer. Shouldn't we just set n_units=[512] and d=128, instead of n_units=[512, 128], d = 128? I may miss something here.

The last question is about the use of GPU. SInce blogcatlog is quite a small graph, I am very surprised that each epoch will take that long(500s). On the same computer, I can process cifar10 at 3000 images/second, so I expect blogcatalog can be trained quite fast. I may misuse your code somehow, can you recall the time for training blogcatalog for each epoch? And how do you set the hyperparameter search range for the larger graph? (For blogcatalog you are using the different network architecture)

Again, thank you very much for all the help and clarification.

**Comment by palash1992** [comment]
Yes, for this dataset I was unable to reproduce the original results by the paper. If you are able to find a better configuration please let me know.

You can use [512] and d=128 too. It is just a slightly shallower network with lesser parameters. As I was testing lower dimensions like 2, 4, 8, 16, 32 and 64 too, I kept it as [512, 128] which can also work with d=128.

I don't recall exactly, but each epoch was much faster than 500 seconds like a minute or something. I can try to run it again next week and get a better estimate. For a larger graph, I increase the size and number of layers in the neural network mostly doubling or quadrupling the size of neurons in each layer.

**Comment by agoodweathercc** [comment]
Thanks a lot for your answer. I will let you know if I found better architecture. 

How many GPU are you using to train SDNE? I tried to run on youtube dataset but the estimated time is like 60k seconds for a single epoch! Let me know if you have to get the rough time estimate for a single epoch on blog and youtube.

Merry Christmas :-)

**Comment by agoodweathercc** [comment]
Hello,

Are you able to run the experiments again and get the single-epoch training time for blog and youtube?

Thanks!

**Comment by agoodweathercc** [comment]
Hello,

I tried quite hard to speed up the code but didn't succeed. 
For Flickr, single epoch takes 15h!(batch size = 500), for blog, it's about 6-7mins. 

I think the memory bound might be the problem. You implement a batch_generator_sdne for model.fit_generator(). I did some research [here](https://stanford.edu/~shervine/blog/keras-how-to-generate-data-on-the-fly), I think the code seems right. But when I run on Flickr,  gpu usage is pretty low() according to the nvidia-smi. So I think fetching the training batch still takes up most of the time.

btw, the number of parameters for the big dataset is pretty large. Roughly the number of parameter is the number of nodes * number of neuron first layer * 2. So for Flickr, if the first layer is 500 neurons, the total number of parameter is 166M. For youtube, it's 1B parameters. In comparsion, VGG has ~155M parameters. (The original author mentioned in the KDD [talk](https://www.youtube.com/watch?v=SHLWSQkvnL4) at around 18min that on youtube it takes **a week** to train SDNE). What's your experience training SDNE on youtube?

**Comment by palash1992** [comment]
Hi, I use 10 GPUs when training on large data sets and it took me a week to train it. Yes, the number of parameters is enormous making it take so long. But that's the algorithm's weakness and that's why even the authors of SDNE only used one big data set.
